great_expectations/agent/agent.py

```diff
@@ -0,0 +1,201 @@
+
+
+@dataclass(frozen=True)
+class GXAgentConfig:
+    """GXAgent configuration.
+    Attributes:
+        organization_id: GX Cloud organization identifier
+        broker_url: address of broker service
+    """
+
+    organization_id: str
+    broker_url: AmqpDsn
+
+
+class GXAgent:
+    """
+    Run GX in any environment from GX Cloud.
+
+    To start the agent, install Python and great_expectations and run `gx-agent`.
+    The agent loads a DataContext configuration from GX Cloud, and listens for
+    user events triggered from the UI.
+    """
+
+    def __init__(self):
+        print("Initializing GX-Agent")
+        self._config = self._get_config_from_env()
+        print("Loading a DataContext - this might take a moment.")
+        self._context: CloudDataContext = get_context(cloud_mode=True)
+        print("DataContext is ready.")
+
+        # Create a thread pool with a single worker, so we can run long-lived
+        # GX processes and maintain our connection to the broker. Note that
+        # the CloudDataContext cached here is used by the worker, therefore
+        # it isn't safe to increase the number of workers.
+        self._executor = ThreadPoolExecutor(max_workers=1)
+        self._current_task: Optional[Future] = None
+
+    def run(self) -> None:
+        """Open a connection to GX Cloud."""
+
+        print("Opening connection to GX Cloud")
+        self._listen()
+        print("Connection to GX Cloud has been closed.")
+
+    def _listen(self) -> None:
+        """Manage connection lifecycle."""
+        subscriber = None
+        try:
+            client = RabbitMQClient(url=self._config.broker_url)
+            subscriber = Subscriber(client=client)
+            print("GX-Agent is ready.")
+            # Open a blocking connection until encountering a shutdown event
+            subscriber.consume(
+                queue=self._config.organization_id,
+                on_message=self._handle_event_as_thread_enter,
+                wait=1,
+            )
+        except KeyboardInterrupt:
+            print("Received request to shutdown.")
+        except (SubscriberError, ClientError) as e:
+            print("Connection to GX Cloud has encountered an error.")
+            print("Please restart the agent and try your action again.")
+            print(e)
+        finally:
+            self._close_subscriber(subscriber)
+
+    def _handle_event_as_thread_enter(self, event_context: EventContext) -> None:
+        """Schedule _handle_event to run in a thread.
+
+        Callback passed to Subscriber.consume which forwards events to
+        the EventHandler for processing.
+
+        Args:
+            event_context: An Event with related properties and actions.
+        """
+        if self._can_accept_new_task() is not True:
+            # signal to Subscriber that we can't process this message right now
+            return event_context.event_processed(retry=True)
```

Comment:
Without any delay, this will cause the agent to spin, trying to execute the event and sending nacks to the broker. Is there a way we could delay attempting to process the next message somehow?

Maybe this is not something to be concerned with for now.

Reply:
thanks - in conjunction with the logic to check correlation IDs, we should only see the same message max 10 times before nacking it with requeue=False. This logic has also been updated to delay a couple seconds before sending the requeue request